# Weather forecasts from OpenWeatherMap API

API documentation: https://openweathermap.org/api/forecast5?collection=current_forecast 

## Libraries and Constants

In [9]:
import os
import requests
import pandas as pd
from dotenv import load_dotenv

In [10]:
load_dotenv(override=True)
connection_string = os.getenv("CON_STRING")
openweather_key = os.getenv("OPENWEATHER_KEY")
print("Connection:", connection_string)
print("API Key:", openweather_key)

Connection: mysql+pymysql://root:Epityxia2026@127.0.0.1:3306/gans
API Key: aba351679836169313916c712cffdd1c


In [11]:
with open(".env", "w") as f:
    f.write("CON_STRING=mysql+pymysql://root:YOUR_PASSWORD@127.0.0.1:3306/gans\n")
    f.write("OPENWEATHER_KEY=your_api_key_here\n")

In [1]:
print(connection_string)

NameError: name 'connection_string' is not defined

In [13]:
load_dotenv()
connection_string = os.getenv("CON_STRING")
openweather_key = os.getenv("OPENWEATHER_KEY")

In [14]:
cities_df = pd.read_sql("cities", con=connection_string)
cities_df

,city_id,name,country,latitude,longitude
0,1,Berlin,Germany,52.5200,13.405
1,2,Hamburg,Germany,53.5500,10.000
2,3,Munich,Germany,48.1375,11.575


In [17]:
berlin = cities_df.loc[0]

url = "https://api.openweathermap.org/data/2.5/forecast"
params = {"lat": berlin["latitude"], "lon": berlin["longitude"], "appid": openweather_key, "units": "metric"}

response = requests.get(url, params=params)
response_json = response.json()
print(response_json.keys())

dict_keys(['cod', 'message', 'cnt', 'list', 'city'])


In [18]:
print(response_json)

{'cod': '200', 'message': 0, 'cnt': 40, 'list': [{'dt': 1787583600, 'main': {'temp': 20.69, 'feels_like': 20.46, 'temp_min': 20.69, 'temp_max': 22.25, 'pressure': 1021, 'sea_level': 1021, 'grnd_level': 1016, 'humidity': 63, 'temp_kf': -1.56, 'dew_point': 13.4}, 'weather': [{'id': 804, 'main': 'Clouds', 'description': 'overcast clouds', 'icon': '04d'}], 'clouds': {'all': 95}, 'wind': {'speed': 1.43, 'deg': 307, 'gust': 1.68}, 'visibility': 10000, 'pop': 0, 'sys': {'pod': 'd'}, 'dt_txt': '2026-08-24 15:00:00'}, {'dt': 1787594400, 'main': {'temp': 20.22, 'feels_like': 19.84, 'temp_min': 19.29, 'temp_max': 20.22, 'pressure': 1021, 'sea_level': 1021, 'grnd_level': 1014, 'humidity': 59, 'temp_kf': 0.93, 'dew_point': 11.96}, 'weather': [{'id': 803, 'main': 'Clouds', 'description': 'broken clouds', 'icon': '04d'}], 'clouds': {'all': 81}, 'wind': {'speed': 1.52, 'deg': 69, 'gust': 1.76}, 'visibility': 10000, 'pop': 0, 'sys': {'pod': 'd'}, 'dt_txt': '2026-08-24 18:00:00'}, {'dt': 1787605200, 'ma

## One city, one forecast

Start small by calling the API on a single location and exploring the returned data for one time

In [19]:
# select one row of cities_df
berlin = cities_df.loc[0]

url = "https://api.openweathermap.org/data/2.5/forecast"
params = {"lat": berlin["latitude"], "lon" :berlin["longitude"], "appid":openweather_key}

response = requests.get(url, params=params)
response

<Response [200]>

A response code of `200` means a successful request! Let's start exploring.

In [20]:
# transform output, check keys
response_json = response.json()
response_json.keys()

dict_keys(['cod', 'message', 'cnt', 'list', 'city'])

Let's go through the keys one at time to understand the data structure

In [21]:
# looks like the status code
response_json['cod']

'200'

In [22]:
# ???
response_json['message']

0

In [23]:
# one forecast every 3 hours means 8 per day; five days times 8 makes 40 forecasts
response_json['cnt']

40

In [24]:
# looks like the forecasts themselves
response_json['list']

[{'dt': 1787583600,
  'main': {'temp': 293.85,
   'feels_like': 293.62,
   'temp_min': 293.85,
   'temp_max': 295.4,
   'pressure': 1021,
   'sea_level': 1021,
   'grnd_level': 1016,
   'humidity': 63,
   'temp_kf': -1.55,
   'dew_point': 286.56},
  'weather': [{'id': 804,
    'main': 'Clouds',
    'description': 'overcast clouds',
    'icon': '04d'}],
  'clouds': {'all': 95},
  'wind': {'speed': 1.43, 'deg': 307, 'gust': 1.68},
  'visibility': 10000,
  'pop': 0,
  'sys': {'pod': 'd'},
  'dt_txt': '2026-08-24 15:00:00'},
 {'dt': 1787594400,
  'main': {'temp': 293.38,
   'feels_like': 293,
   'temp_min': 292.44,
   'temp_max': 293.38,
   'pressure': 1021,
   'sea_level': 1021,
   'grnd_level': 1014,
   'humidity': 59,
   'temp_kf': 0.94,
   'dew_point': 285.12},
  'weather': [{'id': 803,
    'main': 'Clouds',
    'description': 'broken clouds',
    'icon': '04d'}],
  'clouds': {'all': 81},
  'wind': {'speed': 1.52, 'deg': 69, 'gust': 1.76},
  'visibility': 10000,
  'pop': 0,
  'sys': {'

In [25]:
# matches 'cnt' above!
len(response_json['list'])

40

In [26]:
# we have this info in the 'cities' table
response_json['city']

{'id': 6545310,
 'name': 'Mitte',
 'coord': {'lat': 52.52, 'lon': 13.405},
 'country': 'DE',
 'population': 329078,
 'timezone': 7200,
 'sunrise': 1787544222,
 'sunset': 1787595260}

Time to explore one forecast and see what to keep.

In [27]:
forecast = response_json['list'][0]
forecast

{'dt': 1787583600,
 'main': {'temp': 293.85,
  'feels_like': 293.62,
  'temp_min': 293.85,
  'temp_max': 295.4,
  'pressure': 1021,
  'sea_level': 1021,
  'grnd_level': 1016,
  'humidity': 63,
  'temp_kf': -1.55,
  'dew_point': 286.56},
 'weather': [{'id': 804,
   'main': 'Clouds',
   'description': 'overcast clouds',
   'icon': '04d'}],
 'clouds': {'all': 95},
 'wind': {'speed': 1.43, 'deg': 307, 'gust': 1.68},
 'visibility': 10000,
 'pop': 0,
 'sys': {'pod': 'd'},
 'dt_txt': '2026-08-24 15:00:00'}

* dt looks like some kind of timestamp
* main
    * temp looks crazy! 292 degrees!  (use units=metric instead)
    * feels_like might be more important for people than the actual temperature
    * temp_min/max might be worth keeping? maybe not
    * humidity might affect if people rent scooters or not
* weather looks like the kind of descriptions a weather app would show
* clouds probably not needed
* wind
    * speed or gust could be important
* visibility probably not needed
* pop 'probability of precipitation' seems useful
* rain and snow seem useful
* sys probably not needed
* dt_txt a much better timestamp

In [28]:
forecast['wind']#["feels_like"]

{'speed': 1.43, 'deg': 307, 'gust': 1.68}

In [29]:
forecast_data = {
    "temp": forecast['main']['temp'],
    "feels_like": forecast['main']['feels_like'],
    "humidity_%": forecast['main']['humidity'],
    "wind_speed": forecast['wind']['speed'],
    "wind_gust": forecast['wind']['gust'],
    "precipitation_%": forecast['pop'],
    # # docs say there's 'rain' and 'snow', but it's not here!
    "forecast_time": forecast['dt_txt']
}
forecast_data

{'temp': 293.85,
 'feels_like': 293.62,
 'humidity_%': 63,
 'wind_speed': 1.43,
 'wind_gust': 1.68,
 'precipitation_%': 0,
 'forecast_time': '2026-08-24 15:00:00'}

## One city, all forecasts

Now that we have useful information from one forecast, let's try to extract the same information from the other 39, too.

In [42]:
forecasts = []
for forecast in response_json['list']:
    forecast_data = {
        "temp": forecast['main']['temp'],
        "feels_like": forecast['main']['feels_like'],
        "humidity_%": forecast['main']['humidity'],
        "wind_speed": forecast['wind']['speed'],
        "wind_gust": forecast['wind']['gust'],
        "precipitation_%": forecast['pop'],
        "rain_3h": forecast.get('rain', {}).get('3h', 0),
        "snow_3h": forecast.get('snow', {}).get('3h', 0),
        # docs say there's 'snow', but it's not here!
        "forecast_time": forecast['dt_txt']
    }
    forecasts.append(forecast_data)
forecasts

[{'temp': 25,
  'feels_like': 24.76,
  'humidity_%': 46,
  'wind_speed': 4.41,
  'wind_gust': 4.8,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-24 15:00:00'},
 {'temp': 23.88,
  'feels_like': 23.68,
  'humidity_%': 52,
  'wind_speed': 3.23,
  'wind_gust': 8.4,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-24 18:00:00'},
 {'temp': 18.77,
  'feels_like': 18.45,
  'humidity_%': 67,
  'wind_speed': 3.26,
  'wind_gust': 11.07,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-24 21:00:00'},
 {'temp': 13.76,
  'feels_like': 13.12,
  'humidity_%': 74,
  'wind_speed': 2.91,
  'wind_gust': 8.82,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-25 00:00:00'},
 {'temp': 11.62,
  'feels_like': 11.19,
  'humidity_%': 90,
  'wind_speed': 1.32,
  'wind_gust': 2.27,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-25 03:00:

Oh no! One of the forecasts has no 'rain'!

In [31]:
forecast # the iteration variable stays stuck wherever the loop broke

{'dt': 1787583600,
 'main': {'temp': 293.85,
  'feels_like': 293.62,
  'temp_min': 293.85,
  'temp_max': 295.4,
  'pressure': 1021,
  'sea_level': 1021,
  'grnd_level': 1016,
  'humidity': 63,
  'temp_kf': -1.55,
  'dew_point': 286.56},
 'weather': [{'id': 804,
   'main': 'Clouds',
   'description': 'overcast clouds',
   'icon': '04d'}],
 'clouds': {'all': 95},
 'wind': {'speed': 1.43, 'deg': 307, 'gust': 1.68},
 'visibility': 10000,
 'pop': 0,
 'sys': {'pod': 'd'},
 'dt_txt': '2026-08-24 15:00:00'}

Indeed, there's no 'rain' in this dictionary, just like we saw no 'snow' before. JSON documents are semi-structured; where there's no data, they might leave things out rather than listing a null value. Thankfully, python dictionaries have a method that allow us to retrieve a key that might not be there: `.get()`. We can even set the value to return if the key isn't found.

In [32]:
forecast.get('dt_txt')

'2026-08-24 15:00:00'

In [33]:
forecast.get('rain', {}) # {} is the default value

{}

By returning an empty dictionary as the default, we can chain more searches together.

In [34]:
# returns a reasonable value for "no rain in the last three hours"
forecast.get('rain', {}).get('3h', 0)

0

In [35]:
# works just fine if there has been rain
response_json['list'][11].get('rain', {}).get('3h', 0)

0

We'll adjust our loop to use `.get()`

In [36]:
forecasts = []
for forecast in response_json['list']:
    forecast_data = {
        "temp": forecast['main']['temp'],
        "feels_like": forecast['main']['feels_like'],
        "humidity_%": forecast['main']['humidity'],
        "wind_speed": forecast['wind']['speed'],
        "wind_gust": forecast['wind']['gust'],
        "precipitation_%": forecast['pop'],
        "rain_3h": forecast.get('rain', {}).get('3h', 0),
        "snow_3h": forecast.get('snow', {}).get('3h', 0),
        "forecast_time": forecast['dt_txt']
    }
    forecasts.append(forecast_data)
forecasts

[{'temp': 293.85,
  'feels_like': 293.62,
  'humidity_%': 63,
  'wind_speed': 1.43,
  'wind_gust': 1.68,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-24 15:00:00'},
 {'temp': 293.38,
  'feels_like': 293,
  'humidity_%': 59,
  'wind_speed': 1.52,
  'wind_gust': 1.76,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-24 18:00:00'},
 {'temp': 291.61,
  'feels_like': 291.03,
  'humidity_%': 58,
  'wind_speed': 1.45,
  'wind_gust': 1.62,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-24 21:00:00'},
 {'temp': 287.64,
  'feels_like': 286.76,
  'humidity_%': 62,
  'wind_speed': 1.83,
  'wind_gust': 2.74,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-08-25 00:00:00'},
 {'temp': 286.87,
  'feels_like': 286.05,
  'humidity_%': 67,
  'wind_speed': 2.05,
  'wind_gust': 3.68,
  'precipitation_%': 0,
  'rain_3h': 0,
  'snow_3h': 0,
  'forecast_time': '2026-0

In [37]:
pd.DataFrame(forecasts)

,temp,feels_like,humidity_%,wind_speed,wind_gust,precipitation_%,rain_3h,snow_3h,forecast_time
0,293.85,293.62,63,1.43,1.68,0.00,0.00,0,2026-08-24 15:00:00
1,293.38,293.00,59,1.52,1.76,0.00,0.00,0,2026-08-24 18:00:00
2,291.61,291.03,58,1.45,1.62,0.00,0.00,0,2026-08-24 21:00:00
3,287.64,286.76,62,1.83,2.74,0.00,0.00,0,2026-08-25 00:00:00
4,286.87,286.05,67,2.05,3.68,0.00,0.00,0,2026-08-25 03:00:00
5,289.73,289.11,64,2.82,5.19,0.00,0.00,0,2026-08-25 06:00:00
6,295.16,294.51,42,4.08,6.32,0.00,0.00,0,2026-08-25 09:00:00
7,298.15,297.52,31,4.21,4.92,0.00,0.00,0,2026-08-25 12:00:00
8,298.50,297.85,29,4.41,4.91,0.00,0.00,0,2026-08-25 15:00:00
9,294.69,293.89,38,3.05,6.82,0.00,0.00,0,2026-08-25 18:00:00


## All cities, all forecasts

Now let's expand one more time by adding a loop to handle each of our three cities.

In [38]:
cities_df

,city_id,name,country,latitude,longitude
0,1,Berlin,Germany,52.5200,13.405
1,2,Hamburg,Germany,53.5500,10.000
2,3,Munich,Germany,48.1375,11.575


In [46]:
forecasts = []
url = "https://api.openweathermap.org/data/2.5/forecast"

for _, row in cities_df.iterrows():
    params = {
        "lat": row["latitude"],
        "lon": row["longitude"],
        "appid": openweather_key,
        "units": "metric"
    }
    response = requests.get(url, params=params)
    response_json = response.json()

    for forecast in response_json['list']:
        forecast_data = {
            "temp": forecast['main']['temp'],
            "feels_like": forecast['main']['feels_like'],
            "humidity_%": forecast['main']['humidity'],
            "wind_speed": forecast['wind']['speed'],
            "wind_gust": forecast['wind']['gust'],
            "precipitation_%": forecast['pop'],
            "rain_3h": forecast.get('rain', {}).get('3h', 0),
            "snow_3h": forecast.get('snow', {}).get('3h', 0),
            "forecast_time": forecast['dt_txt'],
            "city_id": row["city_id"]
        }
        forecasts.append(forecast_data)

forecast_df = pd.DataFrame(forecasts)
forecast_df

,temp,feels_like,humidity_%,wind_speed,wind_gust,precipitation_%,rain_3h,snow_3h,forecast_time,city_id
0,20.58,20.34,63,1.43,1.68,0.00,0.00,0,2026-08-24 15:00:00,1
1,20.15,19.76,59,1.52,1.76,0.00,0.00,0,2026-08-24 18:00:00,1
2,18.42,17.83,58,1.45,1.62,0.00,0.00,0,2026-08-24 21:00:00,1
3,14.49,13.61,62,1.83,2.74,0.00,0.00,0,2026-08-25 00:00:00,1
4,13.72,12.90,67,2.05,3.68,0.00,0.00,0,2026-08-25 03:00:00,1
...,...,...,...,...,...,...,...,...,...,...
115,14.26,14.20,94,3.08,10.92,1.00,0.27,0,2026-08-29 00:00:00,3
116,13.84,13.66,91,4.06,13.16,0.34,0.16,0,2026-08-29 03:00:00,3
117,14.73,14.43,83,4.00,12.20,0.00,0.00,0,2026-08-29 06:00:00,3
118,20.02,19.44,52,6.40,10.26,0.00,0.00,0,2026-08-29 09:00:00,3


In [47]:
# useful for writing the SQL table definition
forecast_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   temp             120 non-null    float64
 1   feels_like       120 non-null    float64
 2   humidity_%       120 non-null    int64  
 3   wind_speed       120 non-null    float64
 4   wind_gust        120 non-null    float64
 5   precipitation_%  120 non-null    float64
 6   rain_3h          120 non-null    float64
 7   snow_3h          120 non-null    int64  
 8   forecast_time    120 non-null    object 
 9   city_id          120 non-null    int64  
dtypes: float64(6), int64(3), object(1)
memory usage: 9.5+ KB


In [48]:
forecast_df.to_sql(
    "forecasts",
    con=connection_string,
    if_exists="append",
    index=False
)

120